crates/accelerate/src/quantum_circuit/circuit_instruction.rs

```diff
@@ -0,0 +1,256 @@
+        self.clone()
+    }
+
+    /// Creates a shallow copy with the given fields replaced.
+    ///
+    /// Returns:
+    ///     CircuitInstruction: A new instance with the given fields replaced.
+    pub fn replace(
+        &self,
+        py: Python<'_>,
+        operation: Option<PyObject>,
+        qubits: Option<&PyAny>,
+        clbits: Option<&PyAny>,
+    ) -> PyResult<Self> {
+        CircuitInstruction::new(
+            py,
+            operation.unwrap_or_else(|| self.operation.clone_ref(py)),
+            Some(qubits.unwrap_or_else(|| self.qubits.as_ref(py))),
+            Some(clbits.unwrap_or_else(|| self.clbits.as_ref(py))),
+        )
+    }
+
+    fn __getstate__(&self, py: Python<'_>) -> PyObject {
+        (
+            self.operation.as_ref(py),
+            self.qubits.as_ref(py),
+            self.clbits.as_ref(py),
+        )
+            .into_py(py)
+    }
+
+    fn __setstate__(&mut self, _py: Python<'_>, state: &PyTuple) -> PyResult<()> {
+        self.operation = state.get_item(0)?.extract()?;
+        self.qubits = state.get_item(1)?.extract()?;
+        self.clbits = state.get_item(2)?.extract()?;
+        Ok(())
+    }
+
+    pub fn __getnewargs__(&self, py: Python<'_>) -> PyResult<PyObject> {
+        Ok((
+            self.operation.as_ref(py),
+            self.qubits.as_ref(py),
+            self.clbits.as_ref(py),
+        )
+            .into_py(py))
+    }
+
+    pub fn __repr__(self_: &PyCell<Self>, py: Python<'_>) -> PyResult<String> {
+        let type_name = self_.get_type().name()?;
+        let r = self_.try_borrow()?;
+        Ok(format!(
+            "{}(\
+            operation={}\
+            , qubits={}\
+            , clbits={}\
+            )",
+            type_name,
+            r.operation.as_ref(py).repr()?,
+            r.qubits.as_ref(py).repr()?,
+            r.clbits.as_ref(py).repr()?
+        ))
+    }
+
+    // Legacy tuple-like interface support.
+    //
+    // For a best attempt at API compatibility during the transition to using this new class, we need
+    // the interface to behave exactly like the old 3-tuple `(inst, qargs, cargs)` if it's treated
+    // like that via unpacking or similar.  That means that the `parameters` field is completely
+    // absent, and the qubits and clbits must be converted to lists.
+    pub fn _legacy_format(&self, py: Python<'_>) -> PyObject {
+        // Note: previously, we converted qubits and clbits to list,
+        //       but this appears no longer necessary.
+        PyTuple::new(
+            py,
+            [
+                self.operation.as_ref(py),
+                self.qubits.as_ref(py),
+                self.clbits.as_ref(py),
+            ],
+        )
+        .into_py(py)
+    }
+
+    pub fn __getitem__(&self, py: Python<'_>, key: &PyAny) -> PyResult<PyObject> {
+        Ok(self
+            ._legacy_format(py)
+            .as_ref(py)
+            .get_item(key)?
+            .into_py(py))
```

Comment:
Yeah, you're right that it would be faster, but I believe the current approach handles negative indices and slicing like you mention, and in theory, this is a straight port of what the original `CircuitInstruction` Python code was doing.

I'm not too keen on finding a way to reuse the index / slice handling code from this file. IMO, it's something that PyO3 should provide out of the box but doesn't. I only implemented it for `CircuitData` because it's widely treated as a `list` throughout Qiskit and is accessed via indexing very often.

Reply:
Sure, I don't think the legacy access pattern is performance critical so we can keep it as is. I agree this is no worse than the status quo in the Python version. It was most just if you thought it was reasonable to implement this it could be made more efficient.